pyrdp/ui/qt.py

```diff
@@ -174,26 +183,40 @@ def setMousePosition(self, x: int, y: int):
         self.mouseY = y
         self.update()
 
+    def scale(self, scale):
+        """
+        Rescale the current widget to a percentage of the height of the RDP session.
+        :param scale: Ex: 0.5 for 50% height and 50% width.
+        """
+        self.scaleValue = scale
+        self.resize(self.sessionWidth, self.sessionHeight)
+
+    def setScaleToWindow(self, status):
+        self.scaleToWindow = status > 0
+
     def resize(self, width: int, height: int):
         """
         Resize widget
         :param width: new width of the widget
         :param height: new height of the widget
         """
-        self._buffer = QImage(width, height, QImage.Format_RGB32)
+        self._buffer = self._buffer.scaled(width, height)
+        self.sessionWidth = width
+        self.sessionHeight = height
         super().resize(width, height)
 
     def paintEvent(self, e: QEvent):
         """
         Call when Qt renderer engine estimate that is needed
         :param e: the event
         """
+        scaleValue = self.scaleValue if self.scaleToWindow else 1
         qp = QPainter(self)
-        qp.drawImage(0, 0, self._buffer)
+        qp.drawImage(0, 0, self._buffer.scaled(self.sessionWidth * scaleValue, self.sessionHeight * scaleValue, aspectMode=Qt.KeepAspectRatio))
```

Comment:
This is correct. This is where the scaling should happen. Also, FYI: a scaling factor is called a `ratio` or factor.

Reply:
right I'll rename